pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -136,16 +136,46 @@ func (t ResolvedPipelineRunTask) IsSuccessful() bool {
 
 // IsFailure returns true only if the run has failed and will not be retried.
 func (t ResolvedPipelineRunTask) IsFailure() bool {
-	if t.IsCustomTask() {
-		return t.Run != nil && t.Run.IsDone() && !t.Run.IsSuccessful()
+	if t.IsCancelled() {
```

Comment:
one weird bit of maybe undefined behavior is that for TaskRuns, if it is marked as cancelled but has status "unknown" (rather than "false"), this is considered not failed and not done. I decided to use the same behavior for Runs as for TaskRuns. If we want to change this behavior I think it belongs in a separate commit.

Reply:
A taskRun is considered as Cancelled if `status` is `false` and `reason` is set to `TaskRunCancelled`. Are you saying the `status` is set to `unknown` when a taskRun is cancelled on demand? I didnt quite follow how thats happening.

Reply:
What I mean is, if a TaskRun has status "unknown" with reason "cancelled", it will be marked as not failed. It's unclear to me whether such a condition can occur and what the reconciler's behavior should be in this scenario. Our tests implicitly depended on this behavior prior to this commit, and now make this behavior explicit.